drivers/i2c/i2c_npcx_controller.c

```diff
@@ -0,0 +1,904 @@
+					data->oper_state, data->port);
+	data->trans_err = -EIO;
+	return data->trans_err;
+}
+
+/* I2C controller isr function */
+static void i2c_ctrl_isr(const struct device *dev)
+{
+	struct smb_fifo_reg *const inst_fifo = HAL_I2C_FIFO_INSTANCE(dev);
+	struct i2c_ctrl_data *const data = DRV_DATA(dev);
+	uint8_t status, tmp;
+
+	status = inst_fifo->SMBST & NPCX_VALID_SMBST_MASK;
+	LOG_DBG("status: %02x, %d", status, data->oper_state);
+
+	/* A 'Bus Error' has been identified */
+	if (IS_BIT_SET(status, NPCX_SMBST_BER)) {
+		/* Generate a STOP condition immediately */
+		i2c_ctrl_stop(dev);
+
+		/* Clear BER Bit */
+		inst_fifo->SMBST |= BIT(NPCX_SMBST_BER);
+
+		/* Make sure slave doesn't hold bus by reading FIFO again */
+		tmp = i2c_ctrl_fifo_read(dev);
+
+		LOG_ERR("Bus error occurred on i2c port%02x!", data->port);
+		data->oper_state = NPCX_I2C_ERROR_RECOVERY;
+
+		/* I/O error occurred */
+		i2c_ctrl_notify(dev, -EIO);
+		return;
+	}
+
+	/* A negative acknowledge has occurred */
+	if (IS_BIT_SET(status, NPCX_SMBST_NEGACK)) {
+		/* Generate a STOP condition immediately */
+		i2c_ctrl_stop(dev);
+
+		/* Clear NEGACK Bit */
+		inst_fifo->SMBST |= BIT(NPCX_SMBST_NEGACK);
+
+		/* End transaction */
+		data->oper_state = NPCX_I2C_WAIT_STOP;
+
+		/* No such device or address */
+		return i2c_ctrl_notify(dev, -ENXIO);
+	}
+
+	/* START, tx FIFO empty or rx FIFO full has occurred */
+	if (IS_BIT_SET(status, NPCX_SMBST_SDAST)) {
+		if (data->is_write) {
+			return i2c_ctrl_handle_write_int_event(dev);
+		} else {
+			return i2c_ctrl_handle_read_int_event(dev);
+		}
+	}
+
+	/* Clear unexpected status bits */
+	inst_fifo->SMBST = status;
+	LOG_ERR("Unexpected  SMBST 0x%02x occurred on i2c port%02x!", status,
+								data->port);
+}
+
+/* NPCX specific I2C controller functions */
+void npcx_i2c_ctrl_mutex_lock(const struct device *i2c_dev)
+{
+	struct i2c_ctrl_data *const data = DRV_DATA(i2c_dev);
+
+	k_sem_take(&data->lock_sem, K_FOREVER);
+}
+
+void npcx_i2c_ctrl_mutex_unlock(const struct device *i2c_dev)
+{
+	struct i2c_ctrl_data *const data = DRV_DATA(i2c_dev);
+
+	k_sem_give(&data->lock_sem);
+}
+
+int npcx_i2c_ctrl_configure(const struct device *i2c_dev, uint32_t dev_config)
+{
+	struct i2c_ctrl_data *const data = DRV_DATA(i2c_dev);
+
+	switch (I2C_SPEED_GET(dev_config)) {
+	case I2C_SPEED_STANDARD:
+		data->bus_freq = NPCX_I2C_BUS_SPEED_100KHZ;
+		break;
+	case I2C_SPEED_FAST:
+		data->bus_freq = NPCX_I2C_BUS_SPEED_400KHZ;
+		break;
+	case I2C_SPEED_FAST_PLUS:
+		data->bus_freq = NPCX_I2C_BUS_SPEED_1MHZ;
+		break;
+	default:
+		return -ERANGE;
+	}
+
+	i2c_ctrl_config_bus_freq(i2c_dev, data->bus_freq);
+	return 0;
+}
+
+int npcx_i2c_ctrl_transfer(const struct device *i2c_dev, struct i2c_msg *msgs,
+			      uint8_t num_msgs, uint16_t addr, int port)
+{
+	struct i2c_ctrl_data *const data = DRV_DATA(i2c_dev);
+	int ret = 0;
+	uint8_t i;
+
+	/* Does bus need recovery? */
+	if (data->oper_state != NPCX_I2C_WRITE_SUSPEND &&
```

Comment:
We should prevent i2c driver entered recovery mode again if the previous i2c_transfer() is called without I2C_MSG_STOP flag.